### PR TITLE
Refresh and type OrderedSet.

### PR DESCRIPTION
**Pull request by jsirois**
Since Python 2.7 has OrderedDict, just use that for the implementation
and add typing support.

### CODE REVIEW

**Comment by jsirois** [comment]
This is taking over for #1164.

File: pex/orderedset.py [hunk 1]
```diff
@@ -10,71 +10,65 @@
...
-class OrderedSet(MutableSet):
-    KEY, PREV, NEXT = range(3)
 
+class OrderedSet(MutableSet, Generic["_I"]):
     def __init__(self, iterable=None):
-        self.end = end = []
-        end += [None, end, end]  # sentinel node for doubly linked list
-        self.map = {}  # key --> [key, prev, next]
+        # type: (Optional[Iterable[_I]]) -> None
+        self.map = OrderedDict()  # type: OrderedDict[_I, None]
```
Comment (stuhood) [inline pex/orderedset.py:27]: If the API is changing anyway (public `self.end` going away), it seems like maybe a good time to hide `self.map`, unless it's consumed within the codebase?

File: pex/orderedset.py [hunk 2]
```diff
@@ -10,71 +10,65 @@
...
-            return "%s()" % (self.__class__.__name__,)
-        return "%s(%r)" % (self.__class__.__name__, list(self))
+            return "{}()".format(
+                self.__class__.__name__,
+            )
+        return "{}({!r})".format(self.__class__.__name__, list(self))
 
     def __eq__(self, other):
         if isinstance(other, OrderedSet):
             return len(self) == len(other) and list(self) == list(other)
```
Comment (stuhood) [inline pex/orderedset.py:73]: Directly comparing `self.map` should be cheaper than converting to a list, as `OrderedDict` It takes ordering into account. Downside might be that Python3's dict does _not_ take ordering into account in eq, so would need a comment to fix that when Python2 support is dropped.

**Review by stuhood — APPROVED** [review]

File: pex/orderedset.py (hunk 1 above)
Comment (jsirois) [reply pex/orderedset.py:27]: SGTM - I changed it back from self._data to self.map at the last minute just to be nice to reviewers.

File: pex/orderedset.py (hunk 2 above)
Comment (jsirois) [reply pex/orderedset.py]: Fixed. I bit the bullet and re-implemented `__eq__` completely taking away its ability to compare `OrderedSet([1, 2, 3]) == {1, 2, 3}` which was all of strange, wrong and unused.

**Pull request merged**